### ISSUE TITLE
Don't run slow clang-tidy checks

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#1337]
`misc-const-correctness` shows an extreme slowdown on large files, causing a >10x regression of clangd reparse time when enabled. Checks are often inadventently enabled e.g. by `misc-*` globs, as much as we wish this weren't the case. I blocked this check in e78165f0ba1e2fbf72b36a36c8560645b69a168a, but a more systematic approach is needed.

Goals:
1. don't let slow checks into releases. We can't assume check authors care about clangd.
2. minimize feature churn between releases
3. minimize maintenance toil
4. document some fairly objective standard of "fast" that users and check authors can understand

Plan:
- only run "fast checks" by default, as an additional filter applied to checks specified in `.clang-tidy` etc
- keep a static "fast check" list, not a "slow check" list, per goal 1.
- maintain this list by benchmarking on a known complex file (suggest `clang/lib/SemaExpr.cpp`) - addresses goal 4
- "fast" means "regresses `ParsedAST::build()`" by less than 10%
- benchmarking is done by a new mode in `clangd --check`, precise within <5%. Script updates the fast list (goal 3).
- avoid noise by using some debounce: add a check to fast list if measurement <8%, remove it if measurement >13%. Addresses goal 2

**Comment by sam-mccall** [comment]
cc @kadircet as we've been discussing this offline

**Comment by florianhumblot** [comment]
've been trying to get some checks to work in clangd such as misc-const-correctness but after some debugging  I found this commit:
https://github.com/llvm/llvm-project/commit/e78165f0ba1e2fbf72b36a36c8560645b69a168a
which seems to disable the check completely.

The main issue seems to be that the slow check is automatically enabled by the glob which can be a surprise to users that don't explicitly check all the options.
Is there a way to get the check moved to a non-glob group or something like that, to allow for users that want the check (and the downside) to opt-in to it?

For now I've made a script that runs clang-tidy externally and posts the results, but it would be nice for the developers to not get surprised by the CI when they submit patches.

**Comment by HighCommander4** [comment]
> The main issue seems to be that the slow check is automatically enabled by the glob which can be a surprise to users that don't explicitly check all the options. Is there a way to get the check moved to a non-glob group or something like that, to allow for users that want the check (and the downside) to opt-in to it?

Perhaps this could take the form of a new `Allowlist` subkey under [ClangTidy](https://clangd.llvm.org/config#clangtidy) in the clangd config file, which does not allow globs, and explicitly enables the listed checks, overriding the performance section of the [built-in blocklist](https://searchfox.org/llvm/rev/314d0dbb2032c9628620a5697dd4215a9b187d27/clang-tools-extra/clangd/TidyProvider.cpp#196).

**Comment by florianhumblot** [comment]
@sam-mccall would you be alright with the suggestion from @HighCommander4 ? 
If so I'll take a stab at implementing this over the weekend.

**Comment by florianhumblot** [comment]
@sam-mccall @kadircet I'd like to pick this up in the next weeks, do you agree with the following plan (from @HighCommander4 )?

 - Add a `AllowList` subkey to the ClangTidy configuration group in the clangd config file
   - Explicitly disallow globs in this key
 - Any check in this `AllowList` will override matching keys in the built-in blocklist
 
I could also add a warning to the clangd logging that lists the blocked keys that were re-enabled to make sure that the users have an easy way to check that the key was really re-enabled. 

I'l wait for your approval for a while before starting :)
Since I'm not too familiar with clangd/llvm, I may be missing something so if you can think of any edge cases or problems with the proposed solution, let me know :)

**Comment by kadircet** [comment]
Sorry for leaving this thread unattended for a long while. We iterated over this with @sam-mccall couple times offline, but forgot to leave any artifacts.

Our final conclusions were towards:
- Only let people to enable slow checks, and never the crashy checks.
- Do this by having a new config option `ClangTidy.EnableSlowChecks` boolean, that defaults to `False`. (if need be we can have an orthogonal option to enable crashing-checks).

The rationale for never enabling crashy checks is, in practice that just causes support requests on clangd side that we can't maintain and degrades experience for average user to the point of having an unusable clangd until they do a through investigation (or ask for support).

This still needs us to have a separate list of "slow" checks, @sam-mccall is going to be working on getting that piece landed this week.

What do you folks think about this?

**Comment by HighCommander4** [comment]
> What do you folks think about this?

This approach is nice is simple, and it sounds like  it will handle the issues that have come up so far, such as users wanting to enable `misc-const-correctness` (which will be in the "slow" list), and users suffering from the hang-prone `bugprone-unchecked-optional-access` (which is being added to the "crashy" list in https://github.com/llvm/llvm-project/pull/69427).

It wouldn't terribly surprise me if we subsequently get reports about users wanting to enable `bugprone-unchecked-optional-access`, and in such cases I tend to lean towards letting users shoot themselves in the foot if they really want to :) But we can leave that discussion to later if/when we get such a request.

The only other thing that occurs to me is: might users want to enable e.g. only one check on the "slow" list that they really care about, and not all of them? That would be facilitated by the more granular `AllowList` configuration. But that can also be considered later, I think it's fine to keep things simple for now.

**Comment by kadircet** [comment]
> The only other thing that occurs to me is: might users want to enable e.g. only one check on the "slow" list that they really care about, and not all of them? That would be facilitated by the more granular AllowList configuration. But that can also be considered later, I think it's fine to keep things simple for now.

We pick up rest of the configuration from check-list provided through regular means. I think if the users already got their hands dirty with clangd-config, they can also Add/Remove tidy checks while at it. That way we can avoid yet-another-way to mention checks :D

**Comment by HighCommander4** [comment]
Ah, good point, they can use `EnableSlowChecks: true` while disabling the slow checks with `ClangTidy: Remove:`. Seems like that use case is covered then!

**Comment by i-ky** [comment]
> Ah, good point, they can use `EnableSlowChecks: true` while disabling the slow checks with `ClangTidy: Remove:`. Seems like that use case is covered then!

How will user know which checks are considered slow/crashy? Figuring them out and removing one by one does not sound very fun. I wish there was a "collective alias" for them, so I could do something like:
```yaml
Diagnostics:
  ClangTidy:
    Add:
    - '*'
    Remove:
    - slow
    - crashy
```
It would make sense to support these aliases on clang-tidy side as well, would be handy for .clang-tidy configs.

**Comment by FalcoGer** [comment]
https://github.com/clangd/clangd/issues/1337#issuecomment-1770214885

> Ah, good point, they can use `EnableSlowChecks: true` while disabling the ones they don't want with `ClangTidy: Remove:`. Seems like that use case is covered then!

That would require users to have a list available of which checks are slow. Maybe the documentation in https://clang.llvm.org/extra/clang-tidy/checks/list.html could have an extra column added for Good, Slow or Unstable (crashes)

**Comment by aminya** [comment]
I run clang-tidy on cpp files manually inside CI, but running it is very slow. I'd like to know which checks are slowing down the check. Is there a list or way to do that?

**Comment by aminya** [comment]
@5chmidti `--enable-check-profile` is very helpful! I noticed that `cert-dcl37-c` and `cert-dcl51-cpp`  as the alises of ` `bugprone-reserved-identifier` is also pretty slow. I disabled it for clang-tidy, but added it back for `clangd`.


.clangd
```yaml
Diagnostics:
  ClangTidy:
    Add: bugprone-reserved-identifier
```

**Comment by aminya** [comment]
I wonder if clang-tidy automatically deduplicates the fixes before checking as it seems the aliases are checked twice! They appear twice in `--enable-check-profile`

**Comment by geza-herman** [comment]
@sam-mccall If clangd has `FastCheckFilter` now, shouldn't `-misc-const-correctness` be removed from `disableUnusableChecks`?

**Comment by HighCommander4** [comment]
@dpar39 could you file a new issue for that please? It's less likely to be forgotten about that way, than a comment on a closed issue.

**Comment by geza-herman** [comment]
I already opened one, though not in this repo, but in llvm-project: https://github.com/llvm/llvm-project/issues/89758 (not sure how these two repos are related)

**Comment by HighCommander4** [comment]
> I already opened one, though not in this repo, but in llvm-project: [llvm/llvm-project#89758](https://github.com/llvm/llvm-project/issues/89758) (not sure how these two repos are related)

Thanks, that works.

(Either issue tracker is fine. This one predates the entire LLVM project moving from Bugzilla to Github.)
